### PR TITLE
Faster recovery from simulated disurptions

### PR DESCRIPTION
**Pull request by bleskes**
In testing infra, one can simulate node GCs, network issues and other problems by adding a disruption to the test cluster. Those disruption are automatically removed after the test is done. At the moment each disruption indicates how long it will take the cluster to heal once the disruption is removed and the test cluster waits for this amount of time. However, more often than not this is an upper bound, causing a much longer wait than needed.  Instead we should push the responsibility of healing to the disruption it self, where we can be smarter about what we wait for.

### CODE REVIEW

**Comment by bleskes** [comment]
@martijnvg can you take a look?

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java [hunk 1]
```diff
@@ -1015,6 +1015,14 @@ private void ensureStableCluster(int nodeCount, TimeValue timeValue, @Nullable S
                     + stateResponse.getState().prettyPrint());
         }
         assertThat(clusterHealthResponse.isTimedOut(), is(false));
+        if (validateOtherNodes) {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java]: I think this good regardless of this change?

File: core/src/test/java/org/elasticsearch/test/InternalTestCluster.java [hunk 2]
```diff
@@ -1566,20 +1566,7 @@ public void clearDisruptionScheme() {
         if (activeDisruptionScheme != null) {
             TimeValue expectedHealingTime = activeDisruptionScheme.expectedTimeToHeal();
             logger.info("Clearing active scheme {}, expected healing time {}", activeDisruptionScheme, expectedHealingTime);
-            activeDisruptionScheme.removeFromCluster(this);
-            // We don't what scheme is picked, certain schemes don't partition the cluster, but process slow, so we need
-            // to to sleep, cluster health alone doesn't verify if these schemes have been cleared.
-            if (expectedHealingTime != null && expectedHealingTime.millis() > 0) {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/test/InternalTestCluster.java]: nice :)

File: core/src/test/java/org/elasticsearch/test/disruption/SingleNodeDisruption.java [hunk 3]
```diff
@@ -80,4 +82,10 @@ public synchronized void testClusterClosed() {
         disruptedNode = null;
     }
 
+    protected void ensureNodeCount(InternalTestCluster cluster) {
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/test/disruption/SingleNodeDisruption.java]: too bad we have to repeat this here again...

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java [hunk 4]
```diff
@@ -579,7 +579,7 @@ public void testMasterNodeGCs() throws Exception {
         // restore GC
         masterNodeDisruption.stopDisrupting();
         ensureStableCluster(3, new TimeValue(DISRUPTION_HEALING_OVERHEAD.millis() + masterNodeDisruption.expectedTimeToHeal().millis()),
-                oldNonMasterNodes.get(0));
+                oldNonMasterNodes.get(0), false);
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java]: why is the `validateOtherNodes` param set to false here?

**Comment by martijnvg** [comment]
@bleskes Left a question, but this change is looking good.

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java]: yeah, I cherry picked this from another place and this came along.. I'll update the pr description.

File: core/src/test/java/org/elasticsearch/test/disruption/SingleNodeDisruption.java (hunk 3 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/test/disruption/SingleNodeDisruption.java]: I felt it didn't merit another base class just for that..

File: core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java (hunk 1 above)
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/DiscoveryWithServiceDisruptionsTests.java]: second thought, I don't like how this is implemented. I'll revert it.

**Comment by bleskes** [comment]
@martijnvg I updated the PR. can you take a second look?

**Comment by martijnvg** [comment]
LGTM

**Pull request closed without merging**